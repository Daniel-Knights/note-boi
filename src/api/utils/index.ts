export * from './decorators';
export * from './error';
export * from './http';
