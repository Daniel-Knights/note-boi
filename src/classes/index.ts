export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
