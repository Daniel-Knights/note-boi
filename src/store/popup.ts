import { ref } from 'vue';

export enum PopupType {
  Auth,
  Error,
  Info,
  ChangePassword,
}

export const openedPopup = ref<PopupType>();
