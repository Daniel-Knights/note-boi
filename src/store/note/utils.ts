import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
