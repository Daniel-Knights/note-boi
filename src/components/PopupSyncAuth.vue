<template>
  <Popup @close="emit('close')">
    <div id="sync-auth">
      <h2 data-test-id="heading">{{ state.isLogin ? 'Login' : 'Signup' }}</h2>
      <form @submit.prevent="handleSubmit" class="sync-auth__form" data-test-id="form">
        <input
          v-model="state.username"
          @input="validateFields.username"
          :class="{ 'sync-auth__input--error': !validation.username }"
          type="text"
          placeholder="Username"
          ref="usernameInput"
        />
        <input
          v-model="state.password"
          @input="validateFields.password"
          :class="{ 'sync-auth__input--error': !validation.password }"
          type="password"
          placeholder="Password"
          data-test-id="password"
        />
        <input
          v-if="!state.isLogin"
          v-model="confirmPassword"
          @input="validateFields.confirmPassword"
          :class="{ 'sync-auth__input--error': !validation.confirmPassword }"
          type="password"
          placeholder="Confirm Password"
          data-test-id="confirm-password"
        />
        <p v-if="state.error.type === ErrorType.Auth" class="sync-auth__error">
          {{ state.error.message }}
        </p>
        <input type="submit" value="Submit" class="button--default" />
      </form>
      <button
        @click="
          state.isLogin = !state.isLogin;
          resetError();
        "
        class="sync-auth__switch"
        data-test-id="switch"
      >
        Switch to {{ state.isLogin ? 'signup' : 'login' }}
      </button>
    </div>
  </Popup>
</template>

<script lang="ts" setup>
import { onMounted, reactive, ref } from 'vue';

import { ErrorType, resetError, state, login, signup } from '../store/sync';

import Popup from './Popup.vue';

const emit = defineEmits(['close']);

const usernameInput = ref<HTMLInputElement | null>(null);
const confirmPassword = ref('');

const validation = reactive({
  username: true,
  password: true,
  confirmPassword: true,
});

const validateFields = {
  username() {
    validation.username = !!state.username;
  },
  password() {
    validation.password = !!state.password;
  },
  confirmPassword() {
    if (!state.isLogin) validation.confirmPassword = !!confirmPassword.value;
  },
};

async function handleSubmit() {
  validateFields.username();
  validateFields.password();
  validateFields.confirmPassword();

  if (!validation.username || !validation.password) {
    return;
  }

  if (state.isLogin) {
    await login();
  } else {
    if (!confirmPassword.value) {
      validation.confirmPassword = false;
      return;
    }
    if (confirmPassword.value !== state.password) {
      state.error = { type: ErrorType.Auth, message: "Passwords don't match" };
      return;
    }

    await signup();
  }

  if (state.error.type === ErrorType.None) {
    confirmPassword.value = '';
    emit('close');
  }
}

onMounted(() => {
  usernameInput.value?.focus();
});
</script>

<style lang="scss" scoped>
#sync-auth,
.sync-auth__form {
  > * + * {
    margin-top: 12px;
  }
}

.sync-auth__form {
  @include v.flex-y(false, center);
}

.sync-auth__input--error {
  border-color: var(--colour__error);
}

.sync-auth__error {
  color: var(--colour__error);
}

.sync-auth__switch {
  cursor: pointer;
  font-size: 14px;
  text-decoration: underline;

  &:hover {
    text-decoration: none;
  }
}
</style>
