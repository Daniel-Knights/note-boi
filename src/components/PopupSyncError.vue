<template>
  <Popup @close="emit('close')">
    <div class="sync-error__message" data-test-id="error-message">
      Error: {{ state.error.message }}
    </div>
    <button @click="tryAgain" class="sync-error__button" data-test-id="try-again">
      Try again?
    </button>
  </Popup>
</template>

<script lang="ts" setup>
import { ErrorType, logout, pull, push, resetError, state } from '../store/sync';

import Popup from './Popup.vue';

const emit = defineEmits(['close']);

function tryAgain() {
  switch (state.error.type) {
    case ErrorType.Push:
      push();
      break;
    case ErrorType.Pull:
      pull();
      break;
    case ErrorType.Logout:
      logout();
      break;
    // no default
  }

  resetError();
  emit('close');
}
</script>

<style lang="scss" scoped>
.sync-error__message {
  font-size: 20px;
  font-weight: 600;
}

.sync-error__button {
  margin-top: 12px;
  text-decoration: underline;

  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
}
</style>
