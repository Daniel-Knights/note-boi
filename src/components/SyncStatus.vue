<template>
  <div id="sync-status">
    <!-- Loading -->
    <div
      v-if="syncState.isLoading"
      class="sync-status__loading-spinner"
      data-test-id="loading"
    ></div>
    <!-- Error -->
    <button
      v-else-if="isSyncError"
      @click="openedPopup = PopupType.Error"
      class="sync-status__error button"
      title="Sync error"
      data-test-id="error"
    >
      <CloudErrorIcon />
    </button>
    <!-- Sync successful -->
    <div
      v-else-if="syncState.isLoggedIn && syncState.unsyncedNoteIds.size === 0"
      title="Changes synced"
      data-test-id="success"
    >
      <CloudTickIcon />
    </div>
    <!-- Sync ready -->
    <button
      v-else
      @click="handlePopupAuthEvent"
      class="sync-status__sync-button button"
      title="Sync changes"
      data-test-id="sync-button"
    >
      <CloudSyncIcon />
    </button>
  </div>
  <PopupSyncAuth v-if="openedPopup === PopupType.Auth" @close="closeSyncPopup(true)" />
  <PopupSyncError v-if="openedPopup === PopupType.Error" @close="closeSyncPopup" />
</template>

<script lang="ts" setup>
import { computed } from 'vue';

import { openedPopup, PopupType } from '../store/popup';
import { ErrorType, logout, pull, resetError, syncState } from '../store/sync';
import { tauriListen } from '../utils';

import PopupSyncAuth from './PopupSyncAuth.vue';
import PopupSyncError from './PopupSyncError.vue';
import CloudErrorIcon from './svg/CloudErrorIcon.vue';
import CloudSyncIcon from './svg/CloudSyncIcon.vue';
import CloudTickIcon from './svg/CloudTickIcon.vue';

const isSyncError = computed(() => {
  return (
    syncState.error.type === ErrorType.Logout ||
    syncState.error.type === ErrorType.Pull ||
    syncState.error.type === ErrorType.Push
  );
});

function handlePopupAuthEvent() {
  // Prevent bug where event.emit triggers event.listen
  if (!syncState.isLoggedIn) {
    openedPopup.value = PopupType.Auth;
  }
}

function closeSyncPopup(reset?: boolean) {
  openedPopup.value = undefined;

  if (reset) {
    resetError();
  }
}

tauriListen('login', () => {
  syncState.isLogin = true;
  handlePopupAuthEvent();
});
tauriListen('logout', logout);
tauriListen('signup', () => {
  syncState.isLogin = false;
  handlePopupAuthEvent();
});

if (syncState.username) {
  pull();
}
</script>

<style lang="scss" scoped>
#sync-status {
  position: absolute;
  right: v.$utility-menu-right;
  bottom: 12px;

  > * {
    @include v.equal-dimensions(v.$utility-menu-width);
  }
}

.sync-status__loading-spinner {
  display: block;
  border: 5px solid var(--colour__interactive);
  border-bottom-color: var(--colour__secondary);
  border-radius: 50%;
  animation: rotation 1s linear infinite;
}

@keyframes rotation {
  to {
    transform: rotate(360deg);
  }
}

.sync-status__error,
.sync-status__sync-button {
  cursor: pointer;
  pointer-events: all;

  &:hover {
    opacity: 0.8;
  }
}
</style>
