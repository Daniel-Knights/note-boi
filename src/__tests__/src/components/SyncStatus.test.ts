import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';

import * as s from '../../../store/sync';
import { AppError, ERROR_CODE } from '../../../classes';
import { openedPopup, PopupType } from '../../../store/popup';
import { tauriInvoke } from '../../../utils';
import { mockApi } from '../../api';
import {
  assertAppError,
  findByTestId,
  getAppDiv,
  getByTestId,
  getTeleportMountOptions,
  waitUntil,
} from '../../utils';

import PopupSyncAuth from '../../../components/PopupSyncAuth.vue';
import PopupSyncError from '../../../components/PopupSyncError.vue';
import SyncStatus from '../../../components/SyncStatus.vue';

describe('SyncStatus', () => {
  it('Mounts when logged out', () => {
    const { calls } = mockApi();
    const wrapper = mount(SyncStatus);

    assert.isTrue(wrapper.isVisible());
    assert.isFalse(findByTestId(wrapper, 'loading').exists());
    assert.isFalse(findByTestId(wrapper, 'error').exists());
    assert.isFalse(findByTestId(wrapper, 'success').exists());
    assert.isTrue(getByTestId(wrapper, 'sync-button').isVisible());
    assert.strictEqual(calls.size, 3);
    assert.isTrue(calls.listeners.has('login'));
    assert.isTrue(calls.listeners.has('logout'));
    assert.isTrue(calls.listeners.has('signup'));
  });

  it('Mounts when logged in', () => {
    const { calls } = mockApi();
    s.syncState.username = 'd';
    s.syncState.isLoggedIn = true;

    const wrapper = mount(SyncStatus);

    assert.isTrue(wrapper.isVisible());
    assert.isFalse(findByTestId(wrapper, 'loading').exists());
    assert.isFalse(findByTestId(wrapper, 'error').exists());
    assert.isTrue(findByTestId(wrapper, 'success').isVisible());
    assert.isFalse(findByTestId(wrapper, 'sync-button').exists());
    assert.strictEqual(calls.size, 3);
    assert.isTrue(calls.listeners.has('login'));
    assert.isTrue(calls.listeners.has('logout'));
    assert.isTrue(calls.listeners.has('signup'));
  });

  it('Displays loading spinner', async () => {
    mockApi();

    const wrapper = mount(SyncStatus);

    s.syncState.username = 'd';

    await tauriInvoke('set_access_token', {
      username: 'd',
      accessToken: 'test-token',
    });

    s.pull();

    await nextTick();

    assert.isTrue(wrapper.isVisible());
    assert.isTrue(getByTestId(wrapper, 'loading').isVisible());
    assert.isFalse(findByTestId(wrapper, 'error').exists());
    assert.isFalse(findByTestId(wrapper, 'success').exists());
    assert.isFalse(findByTestId(wrapper, 'sync-button').exists());

    await waitUntil(() => !s.syncState.loadingCount);

    assert.isFalse(findByTestId(wrapper, 'loading').exists());
    assert.isFalse(findByTestId(wrapper, 'error').exists());
    assert.isTrue(getByTestId(wrapper, 'success').isVisible());
    assert.isFalse(findByTestId(wrapper, 'sync-button').exists());
  });

  it('Displays error icon and opens popup on click', async () => {
    mockApi();
    s.syncState.appError = new AppError({
      code: ERROR_CODE.PULL,
      display: { sync: true },
      retry: {
        fn: () => {
          //
        },
      },
    });

    const appDiv = getAppDiv();

    document.body.appendChild(appDiv);

    const teleportMountOptions = getTeleportMountOptions(appDiv);
    const wrapper = mount(SyncStatus, teleportMountOptions);

    assert.isTrue(wrapper.isVisible());

    const errorButton = getByTestId(wrapper, 'error');
    await errorButton.trigger('click');

    assert.strictEqual(openedPopup.value, PopupType.Error);
    assert.isTrue(findByTestId(wrapper, 'error-message').exists());
    assert.isTrue(findByTestId(wrapper, 'try-again').exists());
  });

  describe('Handles popups', () => {
    it('PopupSyncError', async () => {
      mockApi({
        request: {
          error: { endpoint: '/login' },
        },
      });

      const appDiv = getAppDiv();

      document.body.appendChild(appDiv);

      const teleportMountOptions = getTeleportMountOptions(appDiv);
      const wrapper = mount(SyncStatus, teleportMountOptions);

      assert.isFalse(findByTestId(wrapper, 'popup-error').exists());

      s.syncState.username = 'd';
      s.syncState.password = '1';

      await s.login();
      await nextTick();
      await findByTestId(wrapper, 'error').trigger('click');

      assert.isTrue(findByTestId(wrapper, 'popup-error').isVisible());
      assert.strictEqual(openedPopup.value, PopupType.Error);

      const resetErrorSpy = vi.spyOn(s, 'resetAppError');

      wrapper.getComponent(PopupSyncError).vm.$emit('close');
      await nextTick();

      expect(resetErrorSpy).not.toHaveBeenCalled();

      assertAppError({
        code: ERROR_CODE.LOGIN,
        retry: { fn: s.login },
        display: { form: true, sync: true },
      });

      assert.isFalse(findByTestId(wrapper, 'popup-error').exists());
      assert.isUndefined(openedPopup.value);
    });

    it('PopupSyncAuth', async () => {
      mockApi();

      const appDiv = getAppDiv();

      document.body.appendChild(appDiv);

      const teleportMountOptions = getTeleportMountOptions(appDiv);
      const wrapper = mount(SyncStatus, teleportMountOptions);
      const wrapperVm = wrapper.vm as unknown as { handlePopupAuthEvent: () => void };

      assert.isTrue(wrapper.isVisible());
      assert.isFalse(findByTestId(wrapper, 'popup-auth').exists());

      wrapperVm.handlePopupAuthEvent();
      await nextTick();

      assert.isTrue(findByTestId(wrapper, 'popup-auth').isVisible());

      wrapper.getComponent(PopupSyncAuth).vm.$emit('close');
      await nextTick();

      assert.isFalse(findByTestId(wrapper, 'popup-auth').exists());
      assert.isUndefined(openedPopup.value);

      s.syncState.isLoggedIn = true;
      wrapperVm.handlePopupAuthEvent();
      await nextTick();

      assert.isFalse(findByTestId(wrapper, 'popup-auth').exists());
      assert.isUndefined(openedPopup.value);
    });
  });
});
