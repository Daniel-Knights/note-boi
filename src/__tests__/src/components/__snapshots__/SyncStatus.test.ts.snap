// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SyncStatus > Displays error icon and opens popup on click > KeyStore.getKey() 1`] = `undefined`;

exports[`SyncStatus > Displays error icon and opens popup on click > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`SyncStatus > Displays error icon and opens popup on click > Storage.get('THEME') 1`] = `null`;

exports[`SyncStatus > Displays error icon and opens popup on click > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`SyncStatus > Displays error icon and opens popup on click > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`SyncStatus > Displays error icon and opens popup on click > Storage.get('USERNAME') 1`] = `null`;

exports[`SyncStatus > Displays error icon and opens popup on click > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
]
`;

exports[`SyncStatus > Displays error icon and opens popup on click > document.body.outerHTML 1`] = `"<body class="theme--system"><div id="app"><div data-v-app=""></div></div></body>"`;

exports[`SyncStatus > Displays error icon and opens popup on click > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`SyncStatus > Displays error icon and opens popup on click > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`SyncStatus > Displays error icon and opens popup on click > openedPopup 1`] = `undefined`;

exports[`SyncStatus > Displays error icon and opens popup on click > selectedTheme 1`] = `"System"`;

exports[`SyncStatus > Displays error icon and opens popup on click > syncState 1`] = `
{
  "appError": AppError {
    "code": 6,
    "display": {
      "sync": true,
    },
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": {
      "fn": [Function],
    },
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`SyncStatus > Displays error icon and opens popup on click > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`SyncStatus > Displays loading spinner > KeyStore.getKey() 1`] = `undefined`;

exports[`SyncStatus > Displays loading spinner > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`SyncStatus > Displays loading spinner > Storage.get('THEME') 1`] = `null`;

exports[`SyncStatus > Displays loading spinner > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`SyncStatus > Displays loading spinner > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`SyncStatus > Displays loading spinner > Storage.get('USERNAME') 1`] = `null`;

exports[`SyncStatus > Displays loading spinner > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "accessToken": "test-token",
        "username": "d",
      },
      "name": "set_access_token",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "username": "d",
      },
      "name": "get_access_token",
    },
  ],
  [
    "request",
    {
      "name": "/notes/pull",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "accessToken": "test-token",
        "username": "d",
      },
      "name": "set_access_token",
    },
  ],
  [
    "emits",
    {
      "calledWith": {
        "data": {
          "is_logged_in": true,
        },
        "isFrontendEmit": true,
      },
      "name": "auth",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "note": {
          "content": {
            "body": "",
            "delta": {
              "ops": [],
            },
            "title": "",
          },
          "id": "id",
          "timestamp": 0,
        },
      },
      "name": "new_note",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "notes": [
          {
            "content": {
              "body": "",
              "delta": {
                "ops": [],
              },
              "title": "",
            },
            "id": "id",
            "timestamp": 0,
          },
        ],
      },
      "name": "sync_local_notes",
    },
  ],
]
`;

exports[`SyncStatus > Displays loading spinner > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`SyncStatus > Displays loading spinner > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`SyncStatus > Displays loading spinner > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [
    {
      "content": {
        "body": "",
        "delta": {
          "ops": [],
        },
        "title": "",
      },
      "id": "id",
      "timestamp": 0,
    },
  ],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`SyncStatus > Displays loading spinner > openedPopup 1`] = `undefined`;

exports[`SyncStatus > Displays loading spinner > selectedTheme 1`] = `"System"`;

exports[`SyncStatus > Displays loading spinner > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": true,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "d",
}
`;

exports[`SyncStatus > Displays loading spinner > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > KeyStore.getKey() 1`] = `undefined`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > Storage.get('THEME') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > Storage.get('USERNAME') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
]
`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > document.body.outerHTML 1`] = `"<body class="theme--system"><div id="app"><div data-v-app=""></div></div></body>"`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > openedPopup 1`] = `undefined`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > selectedTheme 1`] = `"System"`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": true,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`SyncStatus > Handles popups > PopupSyncAuth > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`SyncStatus > Handles popups > PopupSyncError > KeyStore.getKey() 1`] = `undefined`;

exports[`SyncStatus > Handles popups > PopupSyncError > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncError > Storage.get('THEME') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncError > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncError > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncError > Storage.get('USERNAME') 1`] = `null`;

exports[`SyncStatus > Handles popups > PopupSyncError > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
  [
    "request",
    {
      "name": "/login",
    },
  ],
]
`;

exports[`SyncStatus > Handles popups > PopupSyncError > document.body.outerHTML 1`] = `"<body class="theme--system"><div id="app"><div data-v-app=""></div></div></body>"`;

exports[`SyncStatus > Handles popups > PopupSyncError > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`SyncStatus > Handles popups > PopupSyncError > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`SyncStatus > Handles popups > PopupSyncError > openedPopup 1`] = `undefined`;

exports[`SyncStatus > Handles popups > PopupSyncError > selectedTheme 1`] = `"System"`;

exports[`SyncStatus > Handles popups > PopupSyncError > syncState 1`] = `
{
  "appError": AppError {
    "code": 2,
    "display": {
      "form": true,
      "sync": true,
    },
    "message": "Server error",
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": {
      "fn": [Function],
    },
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "1",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "d",
}
`;

exports[`SyncStatus > Handles popups > PopupSyncError > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`SyncStatus > Mounts when logged in > KeyStore.getKey() 1`] = `undefined`;

exports[`SyncStatus > Mounts when logged in > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`SyncStatus > Mounts when logged in > Storage.get('THEME') 1`] = `null`;

exports[`SyncStatus > Mounts when logged in > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`SyncStatus > Mounts when logged in > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`SyncStatus > Mounts when logged in > Storage.get('USERNAME') 1`] = `null`;

exports[`SyncStatus > Mounts when logged in > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
]
`;

exports[`SyncStatus > Mounts when logged in > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`SyncStatus > Mounts when logged in > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`SyncStatus > Mounts when logged in > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`SyncStatus > Mounts when logged in > openedPopup 1`] = `undefined`;

exports[`SyncStatus > Mounts when logged in > selectedTheme 1`] = `"System"`;

exports[`SyncStatus > Mounts when logged in > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": true,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "d",
}
`;

exports[`SyncStatus > Mounts when logged in > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`SyncStatus > Mounts when logged out > KeyStore.getKey() 1`] = `undefined`;

exports[`SyncStatus > Mounts when logged out > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`SyncStatus > Mounts when logged out > Storage.get('THEME') 1`] = `null`;

exports[`SyncStatus > Mounts when logged out > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`SyncStatus > Mounts when logged out > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`SyncStatus > Mounts when logged out > Storage.get('USERNAME') 1`] = `null`;

exports[`SyncStatus > Mounts when logged out > allCalls 1`] = `
[
  [
    "listeners",
    {
      "name": "login",
    },
  ],
  [
    "listeners",
    {
      "name": "logout",
    },
  ],
  [
    "listeners",
    {
      "name": "signup",
    },
  ],
]
`;

exports[`SyncStatus > Mounts when logged out > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`SyncStatus > Mounts when logged out > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`SyncStatus > Mounts when logged out > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`SyncStatus > Mounts when logged out > openedPopup 1`] = `undefined`;

exports[`SyncStatus > Mounts when logged out > selectedTheme 1`] = `"System"`;

exports[`SyncStatus > Mounts when logged out > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`SyncStatus > Mounts when logged out > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
